gateways/server/pulsar/start.go

```diff
@@ -0,0 +1,146 @@
+type EventListener struct {
+	Logger *logrus.Logger
+	// k8sClient is kubernetes client
+	K8sClient kubernetes.Interface
+}
+
+// StartEventSource starts an event source
+func (listener *EventListener) StartEventSource(eventSource *gateways.EventSource, eventStream gateways.Eventing_StartEventSourceServer) error {
+	defer server.Recover(eventSource.Name)
+
+	logrus.WithField(common.LabelEventSource, eventSource.Name).Info(">> StartEventSource")
+
+	log := listener.Logger.WithField(common.LabelEventSource, eventSource.Name)
+	log.Info("started processing the event source...")
+
+	dataCh := make(chan []byte)
+	errorCh := make(chan error)
+	doneCh := make(chan struct{}, 1)
+
+	go listener.listenEvents(eventSource, dataCh, errorCh, doneCh)
+	return server.HandleEventsFromEventSource(eventSource.Name, eventStream, dataCh, errorCh, doneCh, listener.Logger)
+}
+
+// listenEvents fires an event when interval completes and item is processed from queue.
+func (listener *EventListener) listenEvents(eventSource *gateways.EventSource, dataCh chan []byte, errorCh chan error, doneCh chan struct{}) {
+
+	log := listener.Logger.WithField(common.LabelEventSource, eventSource.Name)
+	log.Info(">> listenEvents...") // todo debug
+
+	var pulsarEventSource *v1alpha1.PulsarEventSource
+	if err := yaml.Unmarshal(eventSource.Value, &pulsarEventSource); err != nil {
+		errorCh <- err
+		return
+	}
+
+	pulsarClient, err := pulsar.NewClient(pulsar.ClientOptions{
+		URL:               pulsarEventSource.PulsarConfigUrl,
+		ConnectionTimeout: parseDuration("ConnectionTimeout", pulsarEventSource.PulsarConfigConnectionTimeout, "30s"),
+		OperationTimeout:  parseDuration("OperationTimeout", pulsarEventSource.PulsarConfigOperationTimeout, "30s")})
+	if err != nil {
+		log.WithError(err).Error("Cannot connect to Pulsar")
+		errorCh <- err
+		return
+	} else {
+		log.WithField("url", pulsarEventSource.PulsarConfigUrl).Info("Created Pulsar client")
+	}
+	defer pulsarClient.Close()
+
+	topicLog := log.WithFields(logrus.Fields{
+		"topic":            pulsarEventSource.PulsarConfigTopic,
+		"subscriptionName": pulsarEventSource.PulsarConfigSubscriptionName,
+		"subscriptionType": pulsarEventSource.PulsarConfigSubscriptionType})
+
+	consumer, err := pulsarClient.Subscribe(pulsar.ConsumerOptions{
+		Topic:            pulsarEventSource.PulsarConfigTopic,
+		SubscriptionName: pulsarEventSource.PulsarConfigSubscriptionName,
+		Type:             parseSubscriptionType(pulsarEventSource.PulsarConfigSubscriptionType),
+	})
+	if err != nil {
+		topicLog.WithError(err).Error("Cannot subscribe to Pulsar topic")
+		errorCh <- err
+		return
+	} else {
+		topicLog.Info("Subscribed to Pulsar topic")
+	}
+	defer consumer.Close()
+
+	messages := consumer.Chan()
+
+	log.Info("Starting listening to Pulsar events...")
+
+	for {
+		select {
+		case msg := <-messages:
+			log.WithFields(logrus.Fields{"topic": msg.Topic(), "id": msg.ID()}).Info("Message received") // todo debug
+			dataCh <- msg.Payload()
```

Comment:
Let's define an Event structure for Pulsar gateway at https://github.com/argoproj/argo-events/blob/master/pkg/apis/events/event-data.go. It can look something like,

```
type PulsarEventData struct {
    Topic string `json:topic`
    ID string `json:id`
    Body interface{} `json:payload`
}

```
You can add more fields as you wish. Also, you can take a look at other messaging queues gateway for reference https://github.com/argoproj/argo-events/blob/master/gateways/server/amqp/start.go#L103